pkg/source/adapter/adapter.go

```diff
@@ -303,3 +308,34 @@ func (a *Adapter) InitOffsets(session sarama.ConsumerGroupSession) error {
 	// At this stage the KafkaSource instance is considered Ready (TODO: update KafkaSource status)
 	return nil
 }
+
+// Default retry configuration, 5 retries, exponential backoff with 50ms delay
+func defaultRetryConfig() *kncloudevents.RetryConfig {
+	retryConfig := kncloudevents.NoRetries()
+	retryConfig.CheckRetry = func(ctx context.Context, response *nethttp.Response, err error) (bool, error) {
```

Comment:
could we not use some of the existing `kncloudevents.CheckRetry` functions?

Reply:
None of the existing CheckRetry functions are compliant. However the code here should move to eventing.

Reply:
compliant as in... what?

Reply:
as in the WIP spec linked in this PR description (https://github.com/knative/specs/blob/218f9655c23ca6a35280afc638b2e3c0a5e83f63/specs/eventing/data-plane.md#event-acknowledgement-and-delivery-retry)

Reply:
@matzew are you saying compliance is not a good reason for adding a new implementation or https://github.com/knative/eventing/pull/5285/files is enough?

Reply:
I believe (quick scan only) that the `SelectiveRetry()` implementation in `message_sender.go` does meet the new spec right?  In fact the spec was changed somewhat to align with that implementation (for better/worse ; )

Reply:
I looked at SelectiveRetry and did not quite meet my expectation of when to retry. In particular:
- it retries on any errors. It seems there are cases it should not (like TooManyRedirect, invalid protocol scheme, etc..)
- it keeps retrying even when the context is cancelled

I agree that there shouldn't be a new implementation here and instead `SelectiveRetry()` should be improved.

Reply:
Yeah, sounds like a good plan - the improvements you mentioned above make sense to me!